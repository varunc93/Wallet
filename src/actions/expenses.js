import uuid from 'uuid';

//Action Generators

//Expenses

export const addExpense = ({description = '', text = '', amount = 0, createdAt = 0} = {}) => ({
    type: "ADD_EXPENSE",
    expense: {
        id: uuid(),
        description,
        text,
        amount,
        createdAt
    }
});

export const removeExpense = ({id} = {}) => ({
    type: "REMOVE_EXPENSE",
    id
});

export const editExpense = (id, updates) => ({   //another way of passing id
    type: "EDIT_EXPENSE",
    id,
    updates
});