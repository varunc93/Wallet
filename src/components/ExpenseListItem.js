import React from 'react';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

// const ExpenseListItem = (props) => (
//     <div>
//         <Link to={`/edit/${props.id}`}>
//             <h3>{props.description}</h3>
//         </Link>
//         <p>{props.amount} - {props.createdAt}</p>
//     </div>
// );

//Same as
const ExpenseListItem = ({id, description, amount, createdAt}) => {
    return (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>{numeral(amount/100).format('$0,0.00')}
            -
            {moment(createdAt).format("MMMM Do, YYYY")}</p>
    </div>
)};

//We need to use connect to be able to access dispatch
export default ExpenseListItem;