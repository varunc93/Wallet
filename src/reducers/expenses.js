//Expense Reducer

const expenseReducerDefault = [];

const expenseReducer = (state = expenseReducerDefault, action) => {
    switch(action.type){
        case "ADD_EXPENSE":
            return [...state, action.expense];                //Using spread operator does not change the existing array.
        case "REMOVE_EXPENSE":
            return state.filter(({id}) =>  id !== action.id);
        case "EDIT_EXPENSE":
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default: return state;
    }
};

export default expenseReducer;