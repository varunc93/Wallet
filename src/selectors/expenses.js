import moment from 'moment';
//Selectors

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {  //de-structure filters
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        return textMatch && startDateMatch && endDateMatch;
    }).sort((a,b) => {
        if(sortBy.toLowerCase() === "date")
            return a.createdAt < b.createdAt ? 1 : -1;  //Print the most recent expense first, if a is created first, then print a last.
        else if(sortBy.toLowerCase() === "amount")
            return a.amount - b.amount < 0 ? 1 : -1; //Print higher amount first
        else
            return 0;
    });
};

export default getVisibleExpenses;