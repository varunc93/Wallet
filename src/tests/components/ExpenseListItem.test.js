import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';
import toJSON from 'enzyme-to-json';

test("Return single expense to edit", () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});


