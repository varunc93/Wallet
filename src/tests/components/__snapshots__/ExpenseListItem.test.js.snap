// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Return single expense to edit 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      A
    </h3>
  </Link>
  <p>
    $5.00
    -
    December 31st, 1969
  </p>
</div>
`;
